### ISSUE TITLE
Possible bug in inputs_for when using cast_assoc and custom changesets

### ISSUE THREAD

**Issue by abitdodgy** [phoenixframework/phoenix_ecto#70]
I ran into a possible bug when using `inputs_for` with `cast_assoc` and custom changesets. This is a standard many-to-many association.

When I specify a changeset with a different name (than _changeset_) in `cast_assoc` (for example, `cast_assoc(:foo, with: :bar_changeset)`), the form fails with this error (see stack trace below):

```
no function clause matching in Phoenix.HTML.FormData.Ecto.Changeset.to_changeset/4
```

Here are my models:

```elixir
defmodule App.Organisation do
  schema "organisations" do
    has_many :memberships, App.Membership
    has_many :users, through: [:memberships, :user]
  end

  def changeset(struct, params \\ %{}) do
    struct
    |> cast(params, [:name])
    |> cast_assoc(:memberships, with: :registration_changeset)
  end
end
```

```elixir
defmodule App.Membership do
  schema "organisation_users" do
    belongs_to :organisation, App.Organisation
    belongs_to :user, App.User
  end

  def registration_changeset(struct, params \\ %{}) do
    struct
    |> cast(params, [:role])
    |> cast_assoc(:user, with: :registration_changeset)
  end
end
```

```elixir
defmodule App.User do
  schema "users" do
    has_many :memberships, App.Membership
    has_many :organisations, through: [:memberships, :organisation]
  end

  def registration_changeset(struct, params \\ %{}) do
    struct
    |> cast(params, [:name, :email, :password])
  end
end
```

Pay attention to `cast_assoc(:memberships, with: :registration_changeset)` and `cast_assoc(:user, with: :registration_changeset)` in _Organisation_ and _Membership_ respectively. When I rename my `registration_changeset` functions to `changeset`, and remove the `with: :registration_changeset` from `cast_assoc`, everything works normally.

Here's my view:

```eex
    <%= form_for @changeset, registration_path(@conn, :create), fn f -> %>
      <div class="form-group">
        <%= label f, :name %>
        <%= text_input f, :name, class: "form-control" %>
        <%= error_tag f, :name %>
      </div>

      <div class="form-group">
        <%= inputs_for f, :memberships, fn membership -> %>
          <%= inputs_for membership, :user, fn user -> %>
            <div class="form-group">
              <%= label user, :name %>
              <%= text_input user, :name, class: "form-control" %>
              <%= error_tag user, :name %>
            </div>
          <% end %>
        <% end %>
      </div>
      <%= submit "Create", class: "btn btn-primary" %>
    <% end %>
```

Here's a stack trace:

```
[error] #PID<0.3754.0> running App.Endpoint terminated
Server: localhost:4000 (http)
Request: GET /registrations/new
** (exit) an exception was raised:
    ** (FunctionClauseError) no function clause matching in Phoenix.HTML.FormData.Ecto.Changeset.to_changeset/4
        (phoenix_ecto) lib/phoenix_ecto/html.ex:207: Phoenix.HTML.FormData.Ecto.Changeset.to_changeset(%App.Membership{__meta__: #Ecto.Schema.Metadata<:built, "organisation_users">, id: nil, inserted_at: nil, organisation: #Ecto.Association.NotLoaded<association :organisation is not loaded>, organisation_id: nil, role: nil, updated_at: nil, user: %App.User{__meta__: #Ecto.Schema.Metadata<:built, "users">, email: nil, id: nil, inserted_at: nil, memberships: #Ecto.Association.NotLoaded<association :memberships is not loaded>, name: nil, organisations: #Ecto.Association.NotLoaded<association :organisations is not loaded>, password: nil, password_hash: nil, updated_at: nil}, user_id: nil}, nil, App.Membership, :registration_changeset)
        (phoenix_ecto) lib/phoenix_ecto/html.ex:78: anonymous fn/8 in Phoenix.HTML.FormData.Ecto.Changeset.to_form/4
        (elixir) lib/enum.ex:1623: Enum."-reduce/3-lists^foldl/2-0-"/3
        (phoenix_ecto) lib/phoenix_ecto/html.ex:76: Phoenix.HTML.FormData.Ecto.Changeset.to_form/4
        (phoenix_html) lib/phoenix_html/form.ex:270: Phoenix.HTML.Form.inputs_for/4
        (App) web/templates/registration/new.html.eex:16: anonymous fn/1 in App.RegistrationView.new.html/1
        (phoenix_html) lib/phoenix_html/form.ex:236: Phoenix.HTML.Form.form_for/4
        (App) web/templates/registration/new.html.eex:6: App.RegistrationView."new.html"/1
        (App) web/templates/layout/registration.html.eex:19: App.LayoutView."registration.html"/1
        (phoenix) lib/phoenix/view.ex:335: Phoenix.View.render_to_iodata/3
        (phoenix) lib/phoenix/controller.ex:642: Phoenix.Controller.do_render/4
        (App) web/controllers/registration_controller.ex:1: App.RegistrationController.action/2
        (App) web/controllers/registration_controller.ex:1: App.RegistrationController.phoenix_controller_pipeline/2
        (App) lib/App/endpoint.ex:1: App.Endpoint.instrument/4
        (App) lib/phoenix/router.ex:261: App.Router.dispatch/2
        (App) web/router.ex:1: App.Router.do_call/2
        (App) lib/App/endpoint.ex:1: App.Endpoint.phoenix_pipeline/1
        (App) lib/plug/debugger.ex:123: App.Endpoint."call (overridable 3)"/2
        (App) lib/App/endpoint.ex:1: App.Endpoint.call/2
        (plug) lib/plug/adapters/cowboy/handler.ex:15: Plug.Adapters.Cowboy.Handler.upgrade/4
```

I'm using Phoenix v1.2.1.

**Comment by josevalim** [comment]
`cast_assoc(:memberships, with: :registration_changeset)` is no longer supported, not even in Ecto. So do `cast_assoc(:memberships, with: &Membership.registration_changeset/2)` or something of sorts.

**Comment by himangshuj** [comment]
Seems like this issue has been resolved in release 4.2
https://github.com/phoenixframework/phoenix_ecto/blob/master/CHANGELOG.md
